### PR TITLE
Equality and diversity carry over bug

### PR DESCRIPTION
**Pull request by tomas-stefano**
## Context

We’re had a provider contact us to say we’re sending old HESA values and this is causing their SRS sync to break.

This is happening because candidates have created Apply accounts in previous years (using previous HESA values) and have carried over into the latest recruitment cycle (which brings over these old values)

After an investigation there are more than 1400 candidates (up to March 22nd)  with outdated HESA codes for the current cycle.

## Changes proposed in this pull request

1. Fix carry over bug around equality and diversity data
2. Add **manual** data migration to fix the current carried over outdated equality and diversity.

## Guidance to review

1. Does carry over from an old cycle (and therefore outdated HESA codes) convert to the most up to date HESA codes?
2. Does the data migrations works?

## Link to Trello card

https://trello.com/c/pRXNcEBK/1197-bug-migrate-carried-over-candidate-hesa-attributes-to-the-latest-values

Please refer to Trello comments to see edge cases and decisions took.

### CODE REVIEW

**Comment by tomas-stefano** [comment]
[Bug: Migrate carried over candidate HESA attributes to the latest values](https://trello.com/c/pRXNcEBK/1197-bug-migrate-carried-over-candidate-hesa-attributes-to-the-latest-values)

**Comment by dcyoung-dev** [comment]
I have concerns when it comes to migrating previous HESA codes from one year to the next where there may be a slight change, i.e. text is the same but the code has changed, or vice-versa. In a similar thought where we set the `disability` to "I do not have any of these disability or health conditions" [based on a `blank` value](https://github.com/DFE-Digital/apply-for-teacher-training/pull/9135/files#diff-a0f38bc556e048e0c74feecf96026412095c45d298586064f06e3372669ca43bR52) are we sure that this is the correct option, could the `blank` represent "PREFER_NOT_TO_SAY" or "NOT_AVAILABLE"?

A preference would be to make the form invalid if we can not find the exact HESA codes for this Recruitment Cycle, allowing the Candidate to manually complete the form.

**Comment by tomas-stefano** [comment]
> I have concerns when it comes to migrating previous HESA codes from one year to the next where there may be a slight change, i.e. text is the same but the code has changed, or vice-versa. In a similar thought where we set the `disability` to "I do not have any of these disability or health conditions" [based on a `blank` value](https://github.com/DFE-Digital/apply-for-teacher-training/pull/9135/files#diff-a0f38bc556e048e0c74feecf96026412095c45d298586064f06e3372669ca43bR52) are we sure that this is the correct option, could the `blank` represent "PREFER_NOT_TO_SAY" or "NOT_AVAILABLE"?
> 
> A preference would be to make the form invalid if we can not find the exact HESA codes for this Recruitment Cycle, allowing the Candidate to manually complete the form.

Good! Yes, I agree. Elliot and I talked today about this edge cases. That's what's the conclusion:

1. Prefer not say on HESA sex will continue to be null
2. Disabilities when carry over empty we set the E&D section as incomplete and force the candidate to answer the question again.
3. Ethnicities on carry over the same above, if is old cycle we set the ethnicity as nil and mark the section incomplete force the candidate to answer the question again.
4. Make sure E&D section should not be marked as complete when we have all the necessary info of all the fields.
5. When we migrate (run the data migration for the wrong data that got leaked into the current cycle) on ethnicities, we only touch on the hesa_code for ethnicities and leave the ethnic_background value as it is (because it is delicate to change now).

File: spec/services/data_migrations/incorrect_equality_and_diversity_migration_spec.rb [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        )
+        expect(@blind.reload.equality_and_diversity['hesa_disabilities']).to eq(['58'])
+        expect(@blind.equality_and_diversity['disabilities']).to eq(
+          ['Blindness or a visual impairment not corrected by glasses'],
+        )
+      end
+    end
+
+    context 'when outdated hesa ethnicity' do
+      before do
```
Comment (tomas-stefano) [inline spec/services/data_migrations/incorrect_equality_and_diversity_migration_spec.rb:148]: TODO: break this into contexts.

File: spec/services/duplicate_application_spec.rb [hunk 2]
```diff
@@ -22,94 +22,415 @@
...
+          { hesa_code: '34', expected_conversion: '101', ethnic_background: 'Chinese' },
+          { hesa_code: '39', expected_conversion: '119', ethnic_background: 'Other Asian background' },
+          { hesa_code: '41', expected_conversion: '142', ethnic_background: 'Mixed - White and Black Caribbean' },
+          { hesa_code: '42', expected_conversion: '141', ethnic_background: 'Mixed - White and Black African' },
+          { hesa_code: '43', expected_conversion: '140', ethnic_background: 'Mixed - White and Asian' },
+          { hesa_code: '49', expected_conversion: '159', ethnic_background: 'Other Mixed background' },
+          { hesa_code: '50', expected_conversion: '180', ethnic_background: 'Arab' },
+          { hesa_code: '80', expected_conversion: '899', ethnic_background: 'Other Ethnic background' },
+          { hesa_code: '90', expected_conversion: '997', ethnic_background: 'Not known' },
+          { hesa_code: '98', expected_conversion: '998', ethnic_background: 'Prefer not to say' },
```
Comment (tomas-stefano) [inline spec/services/duplicate_application_spec.rb:359]: I moved all this code to the HESA converter. I think I will just test here some edge cases but testing just that the HesaConverter is being called. I didn't made the changes yet. I will do tomorrow.

File: app/services/duplicate_application.rb [hunk 3]
```diff
@@ -70,8 +74,7 @@ def duplicate
         change_references_to_not_requested_yet(references_cancelled_at_eoc)
       end
 
-      references_completed = apply_again? && new_application_form.complete_references_information?
-      new_application_form.update!(references_completed:)
+      new_application_form.update!(references_completed: false)
```
Comment (elceebee) [inline app/services/duplicate_application.rb:77]: Is this related to the HESA work?

Comment (tomas-stefano) [reply app/services/duplicate_application.rb:77]: Actually no. We don't have the concept of apply_2 anymore and no one is calling apply_2 (a.k.a apply_again?) so this will always be false. But I can't undo this because it is not related to HESA.

Comment (tomas-stefano) [reply app/services/duplicate_application.rb:77]: Wdyt?

Comment (elceebee) [reply app/services/duplicate_application.rb:77]: I don't feel strongly about it. Just wasn't sure if I was missing something.

**Review by elceebee — APPROVED** [review]

File: spec/services/duplicate_application_spec.rb [hunk 4]
```diff
@@ -22,94 +24,240 @@
...
+      @original_application_form.update!(
+        equality_and_diversity_completed: true,
+        equality_and_diversity: @original_application_form.equality_and_diversity.merge(
+          disabilities: nil,
+          hesa_disabilities: nil,
+        ),
+      )
+    end
+
+    it 'sets disabilities as empty' do
```
Comment (inulty-dfe) [inline spec/services/duplicate_application_spec.rb:53]: `s/empty/nil/`

File: spec/services/duplicate_application_spec.rb [hunk 5]
```diff
@@ -22,94 +24,240 @@
...
+        equality_and_diversity: @original_application_form.equality_and_diversity.merge(
+          disabilities: [],
+          hesa_disabilities: [],
+        ),
+      )
+    end
 
-      expect(duplicate_application_form.application_references.count).to eq 3
-      expect(duplicate_application_form.application_references).to all(be_feedback_provided.or(be_not_requested_yet))
+    it 'sets disabilities as empty' do
```
Comment (inulty-dfe) [inline spec/services/duplicate_application_spec.rb:73]: `s/empty/nil`

File: spec/services/duplicate_application_spec.rb [hunk 6]
```diff
@@ -22,94 +24,240 @@
...
+      @original_application_form.update!(
+        equality_and_diversity_completed: true,
+        equality_and_diversity: @original_application_form.equality_and_diversity.merge(
+          ethnic_background: nil,
+          hesa_ethnicity: nil,
+        ),
+      )
+    end
+
+    it 'sets the ethnicity as blank' do
```
Comment (inulty-dfe) [inline spec/services/duplicate_application_spec.rb:93]: `s/blank/nil/`

File: spec/services/hesa_converter_spec.rb [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      expect(hesa_converter.hesa_sex).to eq(
+        data[:expected_hesa_sex],
+      )
+      expect(hesa_converter.sex).to eq(
+        data[:expected_sex] || data[:sex],
+      )
+    end
+  end
+
+  # Below I added all scenarios I could caught in the DB from 2020 til 2024
```
Comment (inulty-dfe) [inline spec/services/hesa_converter_spec.rb:26]: > could catch

File: spec/services/hesa_converter_spec.rb [hunk 8]
```diff
@@ -0,0 +1,206 @@
+require 'rails_helper'
+
+RSpec.describe HesaConverter do
+  shared_examples 'convert sex field' do |data|
+    it "converts old hesa codes from #{data[:recruitment_cycle_year]} cycle for '#{data[:sex]}' into the most up to date HESA codes" do
+      recruitment_cycle_year = data[:recruitment_cycle_year]
+      application_form = create(:application_form, :completed)
+      application_form.update!(
```
Comment (inulty-dfe) [inline spec/services/hesa_converter_spec.rb:8]: We can use `build` and `assign_attributes` here to speed up the tests significantly

**Review by inulty-dfe — COMMENTED** [review]
Good work, complicated area!

Few suggestions

File: app/lib/hesa/disability.rb [hunk 9]
```diff
@@ -1,6 +1,7 @@
 module Hesa
   class Disability
     DisabilityStruct = Struct.new(:hesa_code, :value, :uuid)
+    NO_DISABILITY_UUID = 'b14e142a-adfe-4646-af5d-8236b6a5b48d'.freeze
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/lib/hesa/disability.rb:4]: It came from DfE-reference-data. You can check with `Hesa::Disability.all(2024)`. The UUID will be the same for all cycles - as it is the goal of the gem.

File: app/services/duplicate_application.rb [hunk 10]
```diff
@@ -83,6 +87,39 @@ def duplicate
...
+        hesa_sex: hesa_converter.hesa_sex,
+        sex: hesa_converter.sex,
+        hesa_disabilities: hesa_converter.hesa_disabilities,
+        disabilities: hesa_converter.disabilities,
+        hesa_ethnicity: hesa_converter.hesa_ethnicity,
+      ),
+    )
+  rescue StandardError => e
+    Sentry.capture_message(
+      "Could not migrate equality_and_diversity data from application form '#{original_application_form.id}' to the #{@recruitment_cycle_year} cycle. The carried over application had incomplete equality_and_diversity information, requiring the candidate to re-answer the section questions again. Exception caught: #{e.message}",
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/services/duplicate_application.rb:105]: Good idea.

[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/services/duplicate_application.rb:105]: Done

File: app/services/hesa_converter.rb [hunk 11]
```diff
@@ -0,0 +1,87 @@
+class HesaConverter
+  attr_reader :application_form, :recruitment_cycle_year
+  delegate :equality_and_diversity, to: :application_form
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/services/hesa_converter.rb:3]: Yes, they are. Moved to after private.

File: config/initializers/hesa_ethnicities.rb [hunk 12]
```diff
@@ -72,11 +72,12 @@ module HesaEthnicityCollections
...
-  HESA_ETHNICITIES_2023_2024 = HESA_ETHNICITIES_2022_2023
-
   # https://www.hesa.ac.uk/collection/c20053/e/ethnic
   HESA_ETHNICITIES_2020_2021 = HESA_ETHNICITIES_2019_2020
   # Unchanged from 2020-2021
   # https://www.hesa.ac.uk/collection/c21053/e/ethnic
   HESA_ETHNICITIES_2021_2022 = HESA_ETHNICITIES_2020_2021
+
+  HESA_ETHNICITIES_2023_2024 = HESA_ETHNICITIES_2022_2023
+  HESA_ETHNICITIES_2024_2025 = HESA_ETHNICITIES_2022_2023
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply config/initializers/hesa_ethnicities.rb:82]: No, Hesa sex has a fallback to look at the latest available constant. I consider at first to add a fallback here, but then I disregard because it would be challenging and I didn't want to scope creep this PR.

File: spec/services/duplicate_application_spec.rb (hunk 4 above)
Comment (tomas-stefano) [reply spec/services/duplicate_application_spec.rb]: Done

File: spec/services/duplicate_application_spec.rb (hunk 5 above)
Comment (tomas-stefano) [reply spec/services/duplicate_application_spec.rb]: Done

File: spec/services/hesa_converter_spec.rb (hunk 8 above)
Comment (tomas-stefano) [reply spec/services/hesa_converter_spec.rb]: Good idea! Much faster 🚀

File: app/services/data_migrations/incorrect_equality_and_diversity_migration.rb [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    SQL
+
+    def change
+      records.find_each do |application_form|
+        hesa_converter = HesaConverter.new(
+          application_form:,
+          recruitment_cycle_year: RecruitmentCycle.current_year,
+        )
+
+        application_form.update_columns(
```
[the remark this answers is not in the document]
Comment (tomas-stefano) [reply app/services/data_migrations/incorrect_equality_and_diversity_migration.rb:19]: Audit comment is not an attribute available at update_columns 🤔. I am avoiding trying to call update or save because I don't want to trigger any callbacks. I can see on audit gem people has the same sort of issue.

**Review by inulty-dfe — APPROVED** [review]
Great success!
